### PR TITLE
Tweak WithVolume method for usability

### LINKED ISSUE #2378: AzureServiceBus resource does not work because of connection string format
There's a bug in our bicep file that generates the wrong connection string format

```
ArgumentException: The value 'xxx.servicebus.windows.net:443/' is not a well-formed Service Bus fully qualified namespace. (Parameter 'fullyQualifiedNamespace')
Azure.Core.Argument.AssertWellFormedServiceBusNamespace(string argumentValue, string argumentName)
Azure.Messaging.ServiceBus.ServiceBusConnection..ctor(string fullyQualifiedNamespace, TokenCredential credential, ServiceBusClientOptions options)
Azure.Messaging.ServiceBus.ServiceBusConnection.CreateWithCredential<TCredential>(string fullyQualifiedNamespace, TCredential credential, ServiceBusClientOptions options)
Azure.Messaging.ServiceBus.ServiceBusClient..ctor(string fullyQualifiedNamespace, object credential, ServiceBusClientOptions options)
Azure.Messaging.ServiceBus.ServiceBusClient..ctor(string fullyQualifiedNamespace, TokenCredential credential, ServiceBusClientOptions options)
Microsoft.Extensions.Hosting.AspireServiceBusExtensions+MessageBusComponent+<>c__DisplayClass0_0<TBuilder>.<AddClient>b__0(ServiceBusClientOptions options, TokenCredential cred) in AspireServiceBusExtensions.cs
+
                return !string.IsNullOrEmpty(connectionString) ?
```

The bug is [here](https://github.com/dotnet/aspire/blob/29101ae43f04314c0bfd28f6cc10061943e94039/src/Aspire.Hosting.Azure/Bicep/servicebus.bicep#L54) that replace logic should no longer be needed.

This should be fixed in the latest sdk build https://github.com/Azure/azure-sdk-for-net/issues/41785.

@eerhardt can you take this one? Ideally we'd update that bicep file to not do any transformation and just use the latest SDK.

### LINKED ISSUE #2738: Rename 'source' parameter of WithVolume<T> method to 'name'
We should change the `source` parameter of the `WithVolume<T>` method to be `name` to better match how volumes are spoken about in Docker 
https://github.com/dotnet/aspire/blob/2cd04efc763d476d45ae760977a125e4e15e1025/src/Aspire.Hosting/Extensions/ContainerResourceBuilderExtensions.cs#L49

### LINKED ISSUE #2739: Update WithVolume<T> method to support anonymous volumes more clearly
In its current form it's not clear if the `WithVolume<T>` extension method supports anonymous volumes. The `string source` parameter isn't marked as nullable, nor is it on `ContainerMountAnnotation.Source`, however it appears that `ApplicationExecutor` does interpret it being `null` as an instruction to create an anonymous volume: https://github.com/dotnet/aspire/blob/2cd04efc763d476d45ae760977a125e4e15e1025/src/Aspire.Hosting/Dcp/ApplicationExecutor.cs#L768-L769

To make it clearer how to add anonymous volumes, we should consider adding an overload of `WithVolume<T>` that omits the `source` parameter, and/or update the nullability of the parameter (and `ContainerMountAnnotation.Source`) to indicate it's optional.

### PR DESCRIPTION
**Pull request by DamianEdwards**
Fixes #2738
Fixes #2739
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2795)

### CODE REVIEW

File: src/Aspire.Hosting/Extensions/ContainerResourceBuilderExtensions.cs [hunk 1]
```diff
@@ -42,13 +42,26 @@ public static IResourceBuilder<ContainerResource> AddContainer(this IDistributed
...
-    /// <param name="source">The source name of the volume.</param>
-    /// <param name="target">The target path where the file or directory is mounted in the container.</param>
-    /// <param name="isReadOnly">A flag that indicates if this is a read-only mount.</param>
+    /// <param name="name">The name of the volume.</param>
+    /// <param name="target">The target path where the volume is mounted in the container.</param>
+    /// <param name="isReadOnly">A flag that indicates if the volume should be mounted as read-only.</param>
+    /// <returns>The <see cref="IResourceBuilder{T}"/>.</returns>
+    public static IResourceBuilder<T> WithVolume<T>(this IResourceBuilder<T> builder, string name, string target, bool isReadOnly = false) where T : ContainerResource
+    {
+        var annotation = new ContainerMountAnnotation(name, target, ContainerMountType.Named, isReadOnly);
```
Comment (JamesNK) [inline src/Aspire.Hosting/Extensions/ContainerResourceBuilderExtensions.cs:51]: If we make this change, then `ContainerMountType.Named` should be renamed.

Options:
* Rename to `ContainerMountType.Volume`
* Add `ContainerMountType.Anonymous` and use it from this method.

I believe `Named` is passed to DCP. Either DCP should change values it accepts to something that makes more sense, or we leave DCP and just map from our value to the DCP value, i.e. Volume -> Named.

**Comment by JamesNK** [comment]
I originally allowed a null name, but feedback was anonymous volumes weren't supported by DCP. I never tried. Have you tested DCP is happy with a null name?

cc @karolz-ms

File: src/Aspire.Hosting/ApplicationModel/ContainerMountAnnotation.cs [hunk 2]
```diff
@@ -14,11 +14,11 @@ public sealed class ContainerMountAnnotation : IResourceAnnotation
...
     /// </summary>
-    /// <param name="source">The source path of the mount.</param>
+    /// <param name="source">The source path if a bind mount or name if a volume. Can be <c>null</c> if the mount is an anonymous volume.</param>
     /// <param name="target">The target path of the mount.</param>
     /// <param name="type">The type of the mount.</param>
     /// <param name="isReadOnly">A value indicating whether the mount is read-only.</param>
-    public ContainerMountAnnotation(string source, string target, ContainerMountType type, bool isReadOnly)
+    public ContainerMountAnnotation(string? source, string target, ContainerMountType type, bool isReadOnly)
     {
         Source = source;
```
Comment (JamesNK) [inline src/Aspire.Hosting/ApplicationModel/ContainerMountAnnotation.cs:23]: Should there be validation that a source value is required if type is binding mount?

File: src/Aspire.Hosting/Extensions/ContainerResourceBuilderExtensions.cs (hunk 1 above)
Comment (DamianEdwards) [reply src/Aspire.Hosting/Extensions/ContainerResourceBuilderExtensions.cs:51]: DCP models its support for mounts aligned with the lower level `--mount` flag  in Docker, which covers all mount types, while the higher level concepts like [bind mounts](https://docs.docker.com/storage/bind-mounts/) and [volumes](https://docs.docker.com/storage/volumes/) are what we're aligned with for these APIs. 

I think the best approach is your first suggestion, i.e. we update `ContainerMountType.Named` to `ContainerMountType.Volume` and update our mapping to DCP.

File: src/Aspire.Hosting/ApplicationModel/ContainerMountAnnotation.cs (hunk 2 above)
Comment (DamianEdwards) [reply src/Aspire.Hosting/ApplicationModel/ContainerMountAnnotation.cs:23]: Yeah that's a good idea. I'll add that validation here.

**Comment by DamianEdwards** [comment]
> I originally allowed a null name, but feedback was anonymous volumes weren't supported by DCP. I never tried. Have you tested DCP is happy with a null name?

@JamesNK yes I tested it locally and it works

image

image

File: src/Aspire.Hosting/ApplicationModel/ContainerMountAnnotation.cs [hunk 3]
```diff
@@ -14,11 +14,11 @@ public sealed class ContainerMountAnnotation : IResourceAnnotation
     /// <summary>
     /// Instantiates a mount annotation that specifies the source and target paths for a mount.
     /// </summary>
-    /// <param name="source">The source path of the mount.</param>
+    /// <param name="source">The source path if a bind mount or name if a volume. Can be <c>null</c> if the mount is an anonymous volume.</param>
```
Comment (karolz-ms) [inline src/Aspire.Hosting/ApplicationModel/ContainerMountAnnotation.cs:17]: Just FYI, DCP does not support anonymous volumes as of today (you have to give the volume a valid name). We could add that, but that the state as of today. Not sure if you want to hold off on the part of introducing explicit support for anonymous volumes, or do the whole change anyway and document the missing functionality.

Comment (DamianEdwards) [reply src/Aspire.Hosting/ApplicationModel/ContainerMountAnnotation.cs:17]: Umm, I just tried it and it worked?

Comment (karolz-ms) [reply src/Aspire.Hosting/ApplicationModel/ContainerMountAnnotation.cs:17]: OK. If it works, let's proceed ☺️ 

TBH the code in DCP does not really handle the anonymous volume scenario, but I can also see how it happens to work anyway. 😆

Comment (DamianEdwards) [reply src/Aspire.Hosting/ApplicationModel/ContainerMountAnnotation.cs:18]: Yeah I think it's as simple as if you ask for a volume and the source isn't specified you get an anonymous volume 😃

File: src/Aspire.Hosting/ApplicationModel/ContainerMountAnnotation.cs (hunk 2 above)
Comment (DamianEdwards) [reply src/Aspire.Hosting/ApplicationModel/ContainerMountAnnotation.cs:33]: Added, including tests.

File: tests/Aspire.Hosting.Tests/ContainerMountAnnotationTests.cs [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+namespace Aspire.Hosting.Tests;
+
+public class ContainerMountAnnotationTests
+{
+    [Theory]
+    [InlineData(null)]
+    [InlineData("")]
+    public void CtorThrowsArgumentNullExceptionIfSourceIsMissingForBindMount(string? source)
+    {
+#pragma warning disable CA1507 // Use nameof to express symbol names: fals positive here, the parameter name being tested isn't the parameter to the test method
```
Comment (karolz-ms) [inline tests/Aspire.Hosting.Tests/ContainerMountAnnotationTests.cs:15]:
```suggestion
#pragma warning disable CA1507 // Use nameof to express symbol names: false positive here, the parameter name being tested isn't the parameter to the test method
```

**Review by karolz-ms — APPROVED** [review]

**Comment by JamesNK** [comment]
FYI you might need to resolve conflicts from https://github.com/dotnet/aspire/pull/2809

(or I will, depending on which merges first)

**Pull request merged**